arraycontext/impl/pytato/utils.py

```diff
@@ -0,0 +1,117 @@
+from pytato.transform import CopyMapper
+from pytools import UniqueNameGenerator
+
+
+class Placeholder(BasePlaceholder):
+    """
+    An equality-comparable :class:`pytato.Placeholder`.
+    """
+    def __init__(self, name: str,
+                 shape: ShapeType,
+                 dtype: np.dtype,
+                 tags: TagsType = frozenset()
+                 ) -> None:
+        super().__init__(name, shape, dtype, tags)
+
+    def __eq__(self, other) -> bool:
+        return Array.__eq__(self, other)
+
+    def __hash__(self) -> int:
+        return Array.__hash__(self)
+
+
+class SizeParam(BaseSizeParam):
+    """
+    An equality-comparable :class:`pytato.SizeParam`.
+    """
+    def __init__(self,
+                 name: str,
+                 tags: TagsType = frozenset()
+                 ) -> None:
+        super().__init__(name, tags=tags)
+
+    def __eq__(self, other) -> bool:
+        return Array.__eq__(self, other)
+
+    def __hash__(self) -> int:
+        return Array.__hash__(self)
+
+
+class Normalizer(CopyMapper):
+    def __init__(self) -> None:
+        super().__init__()
+        self.bound_arguments: Dict[str, Any] = {}
+        self.vng = UniqueNameGenerator()
+        self.seen_inputs: Set[str] = set()
+
+    def map_data_wrapper(self, expr: DataWrapper) -> Array:
+        if expr.name is not None:
+            if expr.name in self.seen_inputs:
+                raise ValueError("Got multiple inputs with the name"
+                                 f"{expr.name} => Illegal.")
+            self.seen_inputs.add(expr.name)
+
+        name = self.vng("_actx_dw")
+        self.bound_arguments[name] = expr.data
+        return Placeholder(name=name,
+                           shape=tuple(self.rec(s) if isinstance(s, Array) else s
+                                       for s in expr.shape),
+                           dtype=expr.dtype,
+                           tags=expr.tags)
+
+    def map_size_param(self, expr: BaseSizeParam) -> Array:
+        if expr.name in self.seen_inputs:
+            raise ValueError("Got multiple inputs with the name"
+                             f"{expr.name} => Illegal.")
+        assert isinstance(expr.name, str)
+        self.seen_inputs.add(expr.name)
+
+        name = self.vng("_actx_sp")
+        return SizeParam(name, expr.tags)
+
+    def map_placeholder(self, expr: BasePlaceholder) -> Array:
+        raise ValueError("Placeholders cannot appear in Normalizer.")
+
+
+def normalize_pt_expr(expr: Array) -> Tuple[Array,
+                                            Mapping[str, Any]]:
+    """
+    Returns ``(normalized_expr, bound_arguments)``.  *normalized_expr* is a
+    normalized form of *expr*, with all instances of
+    :class:`pytato.DataWrapper` replaced with instances of :class:`Placeholder`
+    in a deterministic manner. The data corresponding to the placeholders in
+    *normalized_expr* is recorded in the mapping *bound_arguments*.
+    """
```

Comment:
Should this function be part of the documentation?

Reply:
I wouldn't expect these functions to be useful in the user-land, so I think it's fine if we don't publicly document them.